[modbus] Deprecate redundant send APIs

exciton opened the pull request:
This PR is part of a series of fixes for modbus and related components.

Core modbus architecture:
- #8032 -- in [2026.3.0](https://esphome.io/changelog/2026.3.0/)
- #15291 in [2026.4.0](https://esphome.io/changelog/2026.4.0/)
- #14172 in [2026.4.0](https://esphome.io/changelog/2026.4.0/)
- #15509 in [2026.5.0](https://esphome.io/changelog/2026.5.0/)
- #11969 in [2026.7.0](https://esphome.io/changelog/2026.7.0/)
- #11987 in [2026.7.0](https://esphome.io/changelog/2026.7.0/)
- #12376 in [2026.7.0](https://esphome.io/changelog/2026.7.0/)
- #17378 in [2026.7.0](https://esphome.io/changelog/2026.7.0/)
- #17282 in [2026.7.0](https://esphome.io/changelog/2026.7.0/)
- #17434 in [2026.7.0](https://esphome.io/changelog/2026.7.0/)
- #17377 (merged)
- #17844 (merged)
- #17846 (merged)
- #17854 (merged)
- #17848 (merged)
- #17859 (merged)
- #17435 (merged)
- #17888 (merged)
- #17886 (merged)
- #17922 (merged)
- #18196 (merge ready)
- #12421 (merge after 11781 & 12376)
- #12614 (merge after 17922)

Overhaul of modbus_controller
- #17677 (merged)
- #11781 (merged)
- #18071 (merge after 11781)
- #18080 (merge after 18071)
- #18082 (merge after 18080)
- #18085 (merge after 18082)

New features for server mode
- #17205 in [2026.7.0](https://esphome.io/changelog/2026.7.0/)
- #17387 (merged)
- #17357 (merged)
- #17264 (merge ready)
- #17464 (merge after 17264)

New features for client mode
- #17467 (merge after 17434, 17387)
- #17676 (merged)
- #18078 (merged)
- #18215 (merge ready)

There are associated tests
- #14395 (tests 8032 above)
- #14845 (tests 11969 above)
- #17345 (tests 11781 above)

---

# Summary 

This is a breaking change for various redundant or poorly designed modbus and modbus_controller APIs. 
- Remove ModbusClientHub::send(address, function_code, start_address, number_of_entitites)
- Remove ModbusClientDevice::send_raw( payload )
- Remove ModbusController::queue_command(ModbusCommandItem)
- Change ModbusCommandItem::create_custom_command to expect no device address (like all other commands)
- Remove ModbusClientHub::ready_for_immediate_send()

## Changes

### Remove ModbusClientHub::send

*Change*
This function is removed entirely.

*Why Change?*
This was a helper function to build and send a subset of standard modbus commands. Sending such commands directly from the ModbusClientHub means that the response is ignored when it returns. This makes no sense at all for read requests (response will go nowhere) and is error prone for write requests. 

There are two alternatives to this:
- Use a ModbusClientDevice instance and call ModbusClientDevice::send which is the same and handles the response properly. Most components use this method already.
- Use ModbusClientHub::send_raw and build the response using one of the helper functions (e.g. create_write_single_pdu). See below for examples.

### Remove ModbusClientDevice::send_raw

*Change*
This function is removed entirely

*Why Change?*
This was a helper function to forward a modbus command (including the device address) to ModbusClientHub. The API is confusing here, since all other send_ commands from the ModbusClientDevice assume the device address stored inside the ModbusClientDevice, whereas send_raw assumed the address was included. 

There are two alternatives to this:
- Use ModbusClientDevice::send_pdu which accepts a modbus PDU vector starting with the function code and data (such as those generated by helper functions).
- Use ModbusClientDevice::send( function_code, start_address, number_of_entities) which is a helper function to generate and send standard functions (register read, coil/input read)


### Remove ModbusController::queue_command

*Change* 
ModbusController::queue_command is removed entirely, along with associated queues it was feeding.
This was used to create one-off modbus commands via modbus_controller (such as write commands within modbus_controller components, or ad-hoc commands configured by lambdas). 

*Why change?*
ModbusCommandItems need to be kept in memory until they are sent and their response is received (or timed out). ```queue_command``` handled this by placing them in a std::list and then cleaning them up when their callback is fired. This is error-prone and unnecessary. The caller should just hold onto the CommandItem (if used), and call send() on it directly.

There is a better option - which is to send directly from modbus (no controller). If you don't care about handling the response, or retrying on failure:

```c++
id(modbus_hub).send_raw({0x01, 0x42, 0xDE, 0xAD, 0xBE, 0xEF});
```

It's also possible to use one of the many helpers (even without a modbus_controller or creating a ModbusDevice class): 

```c++
std::vector<uint8_t> payload = {0x01}; // Address of device
esphome::modbus::helpers::create_write_single_pdu(payload, 0xDEAD, 0xBEEF);
id(client_hub).send_raw(payload);
```

Alternatively the CommandItem method can be used if the command_item is held in memory (e.g. static variable), via the send() method on the CommandItem (no need to queue with controller). This is useful if you want to utilise the retry features of modbus_controller and track failures.

Old style:
```c++
esphome::modbus_controller::ModbusController *controller = id(modbus_controller_a);
std::vector<uint8_t> payload = {0x01, 0x42, 0xDE, 0xAD, 0xBE, 0xEF};
esphome::modbus_controller::ModbusCommandItem command = modbus_controller::ModbusCommandItem::create_custom_command(controller, payload);
controller->queue_command(command);
```
New style:
```c++
esphome::modbus_controller::ModbusController *controller = id(modbus_controller_a);
std::vector<uint8_t> payload = { 0x42, 0xDE, 0xAD, 0xBE, 0xEF }; //Notice device address is removed
static esphome::modbus_controller::ModbusCommandItem command; //Notice static keyword
command = modbus_controller::ModbusCommandItem::create_custom_command(controller, payload);
command.send();
```

This avoids the controller having to hold references to the one-off write commands coming from various modbus sensors that are read/write. Now those sensors just keep hold of the reference to their write command themselves.


### Change ModbusCommandItem::create_custom_command

*Change*
Previously create_custom_command required the caller to build a modbus command including the device address at the start. This change removes that requirement and utilises the device address already stored as part of the modbus_controller instance.

*Why change?*
All other ModbusCommandItem factory functions use the modbus_controller device address, and handling two different API schemes is confusing and error prone. This way all factory functions will generate a Protocol Data Unit (PDU) which includes only the function code and data. The device address will be prepended automatically by the ModbusClientDevice class.

Users of custom_commands will need to remove the device address from the start of their command, and ensure the address matches the one set in modbus_controller (or set up a new controller with the correct address).

### Remove ModbusClientHub::ready_for_immediate_send()

*Change*
One component (growat_solar) used to wait for modbus to be clear for an immediate send before sending a command, and would use the loop to constantly check this. This change removes that function entirely, and modbus devices can just use the queueing functionality of the modbus component to have their command sent in order.

*Why Change?*
It's simpler to just queue and wait than poll for an empty slot to open up on the bus. 

## TODO
```number_to_payload``` and ```payload_to_number``` heap allocating helpers should likely be deprecated and replaced with a buffer api.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [X] Developer breaking change (an API change that could break external components)
- [X] Code quality improvements to existing code or addition of tests
- [ ] Other

**Related issue or feature (if applicable):**

- fixes <link to issue>

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):**

- esphome/esphome-docs#<esphome-docs PR number goes here>

## Test Environment

- [ ] ESP32
- [X] ESP32 IDF
- [ ] ESP8266
- [ ] RP2040
- [ ] BK72xx
- [ ] RTL87xx
- [ ] nRF52840

## Example entry for `config.yaml`:

```yaml
# Example config.yaml

```

## Checklist:
  - [X] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

exciton committed:
Fixes various modbus timing bugs and follows the spec more closely.

The issues:
- Frame delay timeout (3.5 characters from spec) is not honoured, incorrectly merging unrelated bytes
- Turnaround delay (application specific) is not present, allowing tx too soon after rx or tx
- UART buffer (120 bytes by default) causes timeouts for long responses which get broken into pieces
- TX is synchronous (with flush) causing long component loop execution times
- Waiting for response didn't care if response was from correct address
- Waiting for response is used for server mode which cannot receive responses
- waiting_for_response() only blocks when waiting, doesn't block for frame delay or turnaround delay
- Multiple components inherit from ModbusDevice but don't block for waiting_for_response()
-- sdm_meter, havells_solar, kuntze, pzemac, pzemdc, selec_meter

The fixes:
- Adds a standard 3.5 character timeout for receiving frames (per spec)
- Adds a configurable turnaround delay to prevent sending too quickly after a previous message
- Adds an additional receive timeout delay when long messages are received, to workarond UART buffering
- Makes TX asynchronous (with set_timeout) to avoid long loops
- Adds a TX buffer for ModbusDevices which don't pay attention to waiting_for_response().
- Modifies waiting_for_response to ensure send can be immediate (ModbusController will behave the same, TX buffer will never fill up)

exciton committed:
Merge branch 'dev' into modbus_timing_issues

exciton committed:
Fix up whitespace issues

exciton committed:
Merge upstream

exciton committed:
Rename waiting_for_response to waiting_for_response_ per clang-tidy

exciton committed:
Fix up formatting issues (clang-tidy)

exciton committed:
Make logging timestamps more useful

exciton committed:
Format comment

exciton committed:
Remove long_frame_delay - not needed if UART has an RxFIFO=1

exciton committed:
Make response asynchronous to avoid long loop times for big responses

exciton committed:
Set RxFIFOFUll to 1 for Arduino and ESP-IDF libraries

This fixes the timing issues related to long responses (> 120 bytes) by
streaming them as they come in rather than receiving large chunks with
gaps between (which seems like a timeout)

exciton committed:
Don't unexpectedly drop duplicate commands, increase send buffer

Only reason to have a send buffer limit is to prevent a crash if the
attempted send rate is faster than physically possible on the bus, and
the queue grows infinitely big.

exciton committed:
Rename !waiting_for_response() to ready_for_immediate_send() for clarity

exciton committed:
Tidy up formatting

exciton committed:
Make turnaround_time default to 100ms.

Modbus spec docs recommend 100-200ms.

exciton committed:
Try to setRXFIFOFull for ESP8266

exciton committed:
Add get_hw_serial to esp80266 UART

exciton committed:
Revert "Try to setRXFIFOFull for ESP8266"

This reverts commit 182bacc5cdc652c6dce3ff4b749ab0553f39b252.

exciton committed:
Add this-> to class members and member functions.

Add some comments to explain timing logic
Remove logs in server mode that aren't needed
Tweak logs to give more useful information
Make default turnaround time match python default

exciton committed:
Merge branch 'dev' of github.com:esphome/esphome into modbus_timing_issues

exciton committed:
Add rx_full_threshold and rx_timeout to UART config

Require rx_full_threshold for modbus devices
Remove custom hacking the rx_full_threshold from within the modbus
component

exciton committed:
Fix python formatting

exciton committed:
Merge branch 'dev' into modbus_timing_issues

exciton committed:
Fix up component tests by adding rx_full_threshold for esp32s

Also narrow the requirement to just esp32 platforms, since that's all
that's implemented right now

exciton committed:
Restore set_parity call which was accidentally deleted

exciton committed:
Don't force flush when there's no flow control pin, use send offset

exciton committed:
Make rx_full_threshold optional and put in long buffer wait

This method strikes a balance which allows the user to set the
rx_full_threshold however they like while accounting for both big and
small thresholds correctly in timeout calculations.

exciton committed:
Fix python lint

exciton committed:
Revert "Fix up component tests by adding rx_full_threshold for esp32s"

This reverts commit 60d065f83b311fb30c0eee3066ca2fdd48da56e9.

exciton committed:
Merge remote-tracking branch 'upstream/dev' into modbus_timing_issues

exciton committed:
Use std::max for frame_delay calculation

Co-authored-by: brambo123 <52667932+brambo123@users.noreply.github.com>

exciton committed:
Fix typo in frame_delay calc

exciton committed:
Add send_wait_time and turnaround_time to modbus tests

exciton committed:
Revert changes to uart - rely on PR #8103 instead.

exciton committed:
Merge branch 'dev' of github.com:esphome/esphome into modbus_timing_issues

exciton committed:
Reorder declarations to save some bytes of memory in modbus.h

Co-authored-by: Keith Burzinski <kbx81x@gmail.com>

exciton committed:
Add comment explaining use of millis() for correct timestamp comparisons

exciton committed:
Merge branch 'dev' into modbus_timing_issues

exciton committed:
Fix merge issue duplicating code inside parse_modbus_byte.

exciton committed:
Move ModbusFunctionCode to modbus.h and remove magic numbers

exciton committed:
Make log messages more useful, and lift some to warnings.

Ensure that response messages that are not expected do not get returned
to the client

exciton committed:
Clean up logging to remove redundant "Modbus" description

exciton committed:
Refactor response logic to be easier to follow

exciton committed:
Remove defer calls for now since modbus_controller could misbehave.

modbus_controller currently assumes a timeout has occurred if modbus is
ready_for_immediate_send() before a response has been delivered.

This could lead to mistaken double-sends if there's sufficient time
between the modbus loop execution and the modbus_controller loop
execution, since the response is delivered later, via defer.

We can put this back in if modbus_controller is cleaned up to have a
more explicit way to understand if a timeout has occured.

exciton committed:
Merge branch 'dev' of github.com:esphome/esphome into modbus_timing_issues

bdraco committed:
Merge branch 'dev' into modbus_timing_issues

exciton committed:
Add comment to justify std::deque in modbus tx_buffer.

exciton committed:
Merge branch 'modbus_timing_issues' of github.com:exciton/esphome into modbus_timing_issues

exciton committed:
Switch modbus tx_buffer from deque to queue

exciton committed:
Fix parsing for modbus server mode

Previously all frames received in server mode were assumed to be client
commands to that server. If there are multiple servers on the bus then
responses from those other servers were not parsed correctly, resulting
in CRC errors and a string of other errors that follow.

This adds a flag to track client commands and server responses such that
the server can make an educated guess about what message it has
received.

It will attempt to re-parse the message in case CRC fails or a timeout
occurs to try to recover (in case there are two client commands in a row
to the same server).

exciton committed:
Remove disable_crc from modbus - it has no useful purpose.

Now that frame parsing is not completely broken, there should be no need
to suppress CRC messages.

There was never a need to treat messages as good when the CRC fails -
this can only lead to problems.

exciton committed:
Split Modbus class into ModbusClient and ModbusServer

These two modes have vastly different behaviour, and warrant being
separate classes.

Clients don't have to handle other client messages (there can be only
one)
Servers don't have to buffer requests (they can only reply immediately)

There are many optimisations possible once these are separated.

For now modbus_controller is broken for servers, but will be fixed in a
subsequent commit.

exciton committed:
Add modbus_server component to handle server interactions.

There is almost no overlap between the client functions of
modbus_controller and the server functions. So it makes sense to split
it out so they are distinct.

exciton committed:
Use subclassing in ModbusDevice Schema for better backwards compat

exciton committed:
Split ModbusDevice into ModbusServerDevice and ModbusClientDevice

For clients use pointers to return the result instead of an address
lookup.

This enables multiple client device configurations to talk to the same
device and retrieve the results separately. For example if one is
polling quickly and another slowly.

We still use address lookup for servers because it's initiated over the
wire.

Add ModbusDevice compatibility class for external components

exciton committed:
Remove allow_duplicates from modbus_controller.

This really shouldn't be needed at all, and with queuing happening in
modbus, it should go away from modbus_controller.

exciton committed:
Drop duplicates from ModbusClient transmit queue

exciton committed:
Add on_modbus_timeout to ModbusDevice

This provides a callback for various communication failures

exciton committed:
Refactor ModbusController to remove command and response queues

Remove loop() completely from ModbusController

Remove command_throttle from ModbusController - allow Modbus to
handle this with turnaround_time

Add on_modbus_timeout to ModbusDevice to provide a callback for
various communication failures

Send ModbusController commands from CommandItem (make them
ModbusDevices) so we don't need to search for where to respond

ModbusController now sends all commands directly (they are queued in
Modbus)

exciton committed:
Fix up device registration for ModbusServer devices

Limit Modbus frames to 256 characters
Rename on_modbus_timeout to on_modbus_no_response since it's not always
a timeout
Add function to clear transmit queue for an address

exciton committed:
Clear transmit queue when send count reached

Rename send_count to unresponded_send_count - only count failures

exciton committed:
Add function to clear pending Modbus messages

Whenever a device goes offline, or we don't care to communicate with it anymore,
we can delete all pending commands to it.

exciton committed:
Rename modbusdevice to controller (since CommandItems are ModbusDevices)

exciton committed:
Move command count to controller and share for the same address

It makes no sense to have separate fail counters for each command - this
would allow far too many failures before stopping.

It's better to pool the count for the whole module/address and set the
whole device offline/online together.

exciton committed:
Simplify skip_offline_updates using modulo and centralise to controller

exciton committed:
Share modbus_controller helpers with modbus and modbus_server

exciton committed:
Remove unneeded consts from modbus_server

exciton committed:
Remove unneeded consts

exciton committed:
Move shared definitions to common modbus_helpers

Technically all the stuff in modbus_helpers is not used by the modbus
component, but by child components (like modbus_controller and
modbus_server), but it seems like a sensible place to put it.

exciton committed:
Increase modbus default timeouts to reduce many common errors

Longer timeouts for modbus are generally always better if the
application allows it.
Assuming a response has timed out is bad when it comes later -
especially since it's a half duplex channel - so the next command
clashes with the "late" response of the previous command.

Here we're setting the timeout to 2s. This only affects clients that
don't respond. If they respond it has no effect.
We also set turnaround time to 600ms. This allows for slightly less than
100 messages per minute, which is plenty for folks sampling their
devices once per minute.

These values can always be changed by the user, but good defaults avoid
unnecessary bug reports.

exciton committed:
Simplify modbus server send API.

Remove a bunch of unnecessary parameter

exciton committed:
Simplify modbus server send API by removing junk

Also reuse some vectors to save multiple copies

exciton committed:
Start to clean up byte-by-byte parsing

exciton committed:
Fix existing modbus components to work with ModbusClientDevice

exciton committed:
Fix existing modbus components to work with ModbusClientDevice

exciton committed:
Clean up modbus byte parsing significantly

exciton committed:
Only clear parsed bytes when parse succeeds.

There is a situation where ModbusServer switches from server message
parsing to client message parsing (because a server failed to respond)
and the frame length may not match. In this case we don't want to hose
all of the bytes received in the buffer, but just the ones that parsed
correctly.

exciton committed:
Calculate timeout for server reply (adhere to frame delay)

exciton committed:
Move float helpers to modbus_helpers

exciton committed:
Fix re-parsing of server rx_buffer after timeout.

This will only rescue one frame for now.

exciton committed:
Simplify parsing by looking for entire frames at once

Avoids byte-by-byte scanning through (except for custom function codes)
Paves the way for interupt driven parsing (where frames are already
separated)

exciton committed:
Clean up timeout calculation to make it reusable.

Reduce calls to millis() by keeping the last_receive_check timestamp
which can be used to calculate various timeouts (i.e. time since we last
checked for bytes, not the current time).

Simplify loop by having the parser check for the latest timeout value.
Timeout can change throughout the course of parsing as the length of the
buffer reduces - the last few bytes in the buffer could represent a
timeout.

exciton committed:
Add unsupported function codes to parser length calculation

For servers on a bus with other devices, this will avoid parse failures,
even if these function codes aren't supported by this device.

exciton committed:
Decouple ModbusController completely from message handling.

Send all messages directly to their respective sensors
Provide default handlers to avoid a lot of messy duplicated code
Don't store RegisterRanges any more, only store the CommandItems

Remove CUSTOM sensor types from attempting to find overlapping ranges.
Can probably simplify even further. I think there are only a few sensor
types that can actually utilise overlapping ranges anyway.

exciton committed:
Remove redundant payload_to_float

exciton committed:
Send illegal function code exception in server mode when appropriate

exciton committed:
Compiler warning fix

exciton committed:
Split out Modbus controller and server tests

exciton committed:
Add logging messages for each server error response

exciton committed:
Merge remote-tracking branch 'upstream/dev' into modbus_timing_issues

exciton committed:
Fix clang-tidy suggestions

exciton committed:
Fix remaining clang-tidy errors

exciton committed:
Update codeowners.

Marking  @martgras as owner of modbus_server as the owner of the original code from modbus_controller, which was essentially just moved into this new component.

Can change ownership if requested.

exciton committed:
Implement a one-shot queue for CommandItems

For components that utilise modbus_controller the previous pattern was to send the controller a CommandItem into a queue. We replicate that here since we need to keep the CommandItems around in memory until the response if received in order to fire the callbacks.
An alternative would be for components to keep the CommandItems locally and directly use send(). This avoids using the queue.

exciton committed:
Send exception code when payload does not produce valid number

Send correct byte count, regardless of whether the correct number of registers were requested.
Give a warning when requested registers does not match byte count.

exciton committed:
Decouple ModbusController from CommandItem

Ensure CommandItems with no controller don't crash
Remove friend relationship between classes
Priovide factory method to create CommandItem linked to controller
Simplify creation of polling command items with constructors

exciton committed:
Add callback for not_sent, constructor, destructor

We add a callback for ModbusClient on_modbus_not_sent to allow us to track when a command doesn't actually go out (so we can clean it up from the queue)

Add convenience constructors to ModbusClientDevice, ModbusServerDevice
We also add a destructor for ModbusClientDevice that removes any references to itself from the modbus queue (via clear_rx_queue_for_device).
Provide a way to dequeue commands based on the device pointer (not just address)

Reset waiting_for_response as soon as we know we have a valid response.
This prevents unneccesary double clearing of waiting_for_response which can happen in the on_modbus_xx callbacks.

exciton committed:
Fix clang-tidy suggestions

exciton committed:
Add  helper to find register type from function code

exciton committed:
Add  helper to find register type from function code

exciton committed:
Refactor modbus send to utilise reusable helpers to create frames

Remove half-baked send function which combines read/write into one error-prone function
Add a send_pdu function to explictly send a modbus pdu (and have ModbusClientDevice add the address)
Move add_crc_to_payload to be a helper
Add a bunch of create_xxx_pdu functions to help generate modbus pdus

Rename ModbusFunctionCode::CUSTOM to ModbusFunctionCode::INVALID. There's no reason to refer to a "function code" as CUSTOM, since custom function codes fall in a range. This is confusing.

exciton committed:
Remove verbose register  details from CommandItem and store payload

Rather than keep all details of the command separate and re-create the payload each time, just store the whole payload and calculate the register details when needed (which is generally only for logs - and can probably be simplified)

exciton committed:
Minor cleanup

exciton committed:
Rename ModbusClient, to ModbusClientHub (and ModbusServerHub)

For clarity to distinguish between modbus_server::ModbusServer class.

exciton committed:
Refactor modbus send to utilise reusable helpers to create frames

Remove half-baked send function which combines read/write into one error-prone function
Add a send_pdu function to explictly send a modbus pdu (and have ModbusClientDevice add the address)
Move add_crc_to_payload to be a helper
Add a bunch of create_xxx_pdu functions to help generate modbus pdus

Rename ModbusFunctionCode::CUSTOM to ModbusFunctionCode::INVALID. There's no reason to refer to a "function code" as CUSTOM, since custom function codes fall in a range. This is confusing.

exciton committed:
Make device callback optionsl in ModbusClientHub

This enables super simple ad-hoc commands to be sent by calling id(modbus_hub_id).send_raw({0x03, ...}) if you don't care about the response.

exciton committed:
Provide programatic way to allow duplicates in modbus queue

exciton committed:
Recreate on_send callback. It fires when command sent to modbus hub.

If this command is already in the queue, and gets skipped (for being duplicate), this implementation will still fire the callback. Would need more thorough surgery to change this.

exciton committed:
Speed up responsiveness of modbus significantly

For ESP32 devices - wake loop when bytes are received (using built-in event queue)
Directly send in the main loop if delay is small enough (small usleep)
Read bytes all at once instead of one at a time.

Decreases average round-trip for a single register request from 28ms to 8.5ms @ 115200 baud. (Min theoretical is ~5ms).

exciton committed:
Simplify ModbusServerDevice to return response or exception directly.

Since exactly one response is required from a server when it receives a request, and the address and function code need to match the request, there's no need to send them explicitly from the device - so we simply return the payload (or exception code) and the ServerHub will send the correct formatted response.

Now the ModbusServerDevice no longer needs to know about the parent (ServerHub). Much simpler and cleaner.

exciton committed:
Merge branch 'dev' into modbus_overhaul

exciton committed:
Merge branch 'modbus_overhaul_server_mode_improvements' into modbus_overhaul_event_driven_speed_up

exciton committed:
Actually disable loop

exciton committed:
Fix clang-tidy recommendations.

exciton committed:
Protect wake_loop_threadsafe for platforms that don't support

Also tidy up delay logic and use delay instead of usleep for wider compatibility.

exciton committed:
Remove allow_duplicates option and instead requeue commands if they are requested again.

This allows for a one-time requeue of a command without needing to explicitly allow the queue to grow indefinitely. It's more efficient to just put the old command to the back of the queue.

exciton committed:
Add CommandPriorities to ModbusCommandItems so we can prioritise the send queue

exciton committed:
Actually prioritise the transmit queue.

exciton committed:
Plumb through continuous reading from modbus_controller to enable continuous mode

exciton committed:
Use on_command_sent for controller callback, which is more reliable as a signal, and closer in time.

exciton committed:
Deprecate modbus_controller queue_command in favour of directly calling send() on CommandItems

This is a breaking change for folks that have custom code to create one-off modbus commands via modbus_controller.

There is a better option - which is to send directly from modbus (no controller). e.g. id(modbus_hub).send_raw({0x01, 0x42, 0xDE, 0xAD, 0xBE, 0xEF});

Alternatively the CommandItem method can be used if the command_item is held in memory (e.g. static variable), via the send() method on the CommandItem (no need to queue with controller).

This avoids the controller having to hold references to the one-off write commands coming from various modbus sensors that are read/write. Now those sensors just keep hold of the reference to their write command themselves.

The consequence is that multiple write commands to the same address in quick succession will favour the most recent command (prior commands will be dequeued upon destruction). This is fine for these particular sensors.

exciton committed:
Fix write register response

exciton committed:
Ensure duplicate commands don't exist inside waiting_for_response

exciton committed:
Don't requeue continuous commands if they fail

exciton committed:
Ensure only one exception response is sent.

exciton committed:
Remove unnecessary code to satisfy clang-tidy

exciton committed:
Adjust case of ModbusFunctionCode

exciton committed:
Rename continuous const to match text.

exciton committed:
Ensure waiting_for_response is cleared even if no device is present

exciton committed:
Deprecate setting device address in custom_commands

exciton committed:
Deprecate ModbusClientHub::send, ModbusClientDevice::send_raw

exciton committed:
Merge branch 'modbus_overhaul_update_clients' into modbus_overhaul_simplify_client_send

exciton committed:
Deprecate ModbusDevice

exciton committed:
Fix frame length calculation for short custom frames

GHswitt commented:
I'm getting 'task watchdog' resets with this PR when a server goes offline (timeout):

```
[00:50:32.456][W][modbus:040]: Stop waiting for response from 1 2005ms after last send
[00:50:36.948][W][modbus:040]: Stop waiting for response from 3 2006ms after last send
[00:50:37.341][W][modbus:633]: Clearing buffer of 37 bytes - parse failed 392ms after last send
[00:50:37.352][W][modbus:633]: Clearing buffer of 37 bytes - parse failed 402ms after last send
[00:50:37.356][W][modbus:633]: Clearing buffer of 16 bytes - parse failed 407ms after last send
[00:50:37.367][W][modbus:633]: Clearing buffer of 37 bytes - parse failed 417ms after last send
[00:50:42.430]E (86807789) task_wdt: Task watchdog got triggered. The following tasks/users did not reset the watchdog in time:
[00:50:42.430]E (86807789) task_wdt:  - loopTask (CPU 1)
[00:50:42.430]E (86807789) task_wdt: Tasks currently running:
[00:50:42.430]E (86807789) task_wdt: CPU 0: IDLE0
[00:50:42.430]E (86807789) task_wdt: CPU 1: loopTask
[00:50:42.430]E (86807789) task_wdt: Aborting.
[00:50:42.430]E (86807789) task_wdt: Print CPU 1 backtrace




[00:50:42.430]Backtrace: 0x42014ede:0x3fcedf70 0x4200c1ef:0x3fcedf90 0x4200c23a:0x3fcedff0 0x4200b256:0x3fcee030 0x4200bef1:0x3fcee050 0x420b7891:0x3fcee070 0x42014385:0x3fcee090 0x42013cf8:0x3fcee0b0 0x4201bcc2:0x3fcee110 0x42006c26:0x3fcee130
WARNING Found stack trace! Trying to decode it
WARNING Decoded 0x42014ede: esphome::crc16(unsigned char const*, unsigned short, unsigned short, unsigned short, bool, bool) at D:\esphome\esphome-yambms\.esphome\build\yambms/src/esphome/core/helpers.cpp:94
WARNING Decoded 0x4200c1ef: esphome::modbus::Modbus::parse_modbus_server_frame_() at D:\esphome\esphome-yambms\.esphome\build\yambms/src/esphome/components/modbus/modbus.cpp:178
WARNING Decoded 0x4200c23a: esphome::modbus::ModbusClientHub::parse_modbus_frames() at D:\esphome\esphome-yambms\.esphome\build\yambms/src/esphome/components/modbus/modbus.cpp:124
 (inlined by) esphome::modbus::ModbusClientHub::parse_modbus_frames() at D:\esphome\esphome-yambms\.esphome\build\yambms/src/esphome/components/modbus/modbus.cpp:119
WARNING Decoded 0x4200b256: esphome::modbus::Modbus::loop() at D:\esphome\esphome-yambms\.esphome\build\yambms/src/esphome/components/modbus/modbus.cpp:29
WARNING Decoded 0x4200bef1: esphome::modbus::ModbusClientHub::loop() at D:\esphome\esphome-yambms\.esphome\build\yambms/src/esphome/components/modbus/modbus.cpp:34
WARNING Decoded 0x420b7891: esphome::Component::call_loop() at D:\esphome\esphome-yambms\.esphome\build\yambms/src/esphome/core/component.cpp:208
WARNING Decoded 0x42014385: esphome::Component::call() at D:\esphome\esphome-yambms\.esphome\build\yambms/src/esphome/core/component.cpp:263
WARNING Decoded 0x42013cf8: esphome::Application::loop() at D:\esphome\esphome-yambms\.esphome\build\yambms/src/esphome/core/application.cpp:168
WARNING Decoded 0x4201bcc2: loop() at D:\esphome\esphome-yambms\.esphome\build\yambms/packages\yambms/yambms_core.yaml:1998
WARNING Decoded 0x42006c26: esphome::loop_task(void*) at D:\esphome\esphome-yambms\.esphome\build\yambms/src/esphome/components/esp32/core.cpp:62 (discriminator 1)




[00:50:42.430]ELF file SHA256: dcec6c99d

[00:50:42.430]Rebooting...
[00:50:42.430]ESP-ROM:esp32s3-20210327
[00:50:43.669]Build:Mar 27 2021
[00:50:43.669]rst:0xc (RTC_SW_CPU_RST),boot:0x2b (SPI_FAST_FLASH_BOOT)
[00:50:43.669]Saved PC:0x4204e8b0
WARNING Decoded 0x4204e8b0: task_wdt_timeout_handling at C:\Users\ipc\.platformio\packages\framework-espidf\components\esp_system\task_wdt/task_wdt.c:427
 (inlined by) task_wdt_isr at C:\Users\ipc\.platformio\packages\framework-espidf\components\esp_system\task_wdt/task_wdt.c:509      
[00:50:43.669]SPIWP:0xee
[00:50:43.669]mode:DIO, clock div:1
[00:50:43.669]load:0x3fce2820,len:0x15c8
[00:50:43.669]load:0x403c8700,len:0xce4
[00:50:43.669]load:0x403cb700,len:0x2f98
[00:50:43.669]entry 0x403c8918
[00:50:43.669]I (24) boot: ESP-IDF 5.5.2 2nd stage bootloader
[00:50:43.669]I (24) boot: compile time Feb  6 2026 11:01:43
[00:50:43.669]I (24) boot: Multicore bootloader
[00:50:43.669]I (25) boot: chip revision: v0.2
[00:50:43.669]I (25) boot: efuse block revision: v1.3
[00:50:43.669]I (25) boot.esp32s3: Boot SPI Speed : 80MHz
[00:50:43.669]I (25) boot.esp32s3: SPI Mode       : DIO
[00:50:43.669]I (25) boot.esp32s3: SPI Flash Size : 16MB
[00:50:43.669]I (25) boot: Enabling RNG early entropy source...
[00:50:43.669]I (26) boot: Partition Table:
[00:50:43.669]I (26) boot: ## Label            Usage          Type ST Offset   Length
[00:50:43.669]I (26) boot:  0 otadata          OTA data         01 00 00009000 00002000
[00:50:43.669]I (26) boot:  1 phy_init         RF data          01 01 0000b000 00001000
[00:50:43.669]I (27) boot:  2 app0             OTA app          00 10 00010000 007c0000
[00:50:43.669]I (27) boot:  3 app1             OTA app          00 11 007d0000 007c0000
[00:50:43.669]I (28) boot:  4 nvs              WiFi data        01 02 00f90000 0006d000
[00:50:43.669]I (28) boot: End of partition table
[00:50:43.669]I (28) esp_image: segment 0: paddr=007d0020 vaddr=3c0d0020 size=208e8h (133352) map
[00:50:43.669]I (53) esp_image: segment 1: paddr=007f0910 vaddr=3fc96800 size=04d3ch ( 19772) load
[00:50:43.669]I (57) esp_image: segment 2: paddr=007f5654 vaddr=40374000 size=0a9c4h ( 43460) load
[00:50:43.669]I (67) esp_image: segment 3: paddr=00800020 vaddr=42000020 size=c2af0h (797424) map
[00:50:43.669]I (208) esp_image: segment 4: paddr=008c2b18 vaddr=4037e9c4 size=07df0h ( 32240) load
[00:50:43.669]I (216) esp_image: segment 5: paddr=008ca910 vaddr=50000000 size=00020h (    32) load
[00:50:43.669]I (224) boot: Loaded app from partition at offset 0x7d0000
[00:50:43.669]I (224) boot: Disabling RNG early entropy source...
[00:50:43.669][I][logger:121]: Log initialized
```

exciton commented:
@GHswitt - any chance you can get a log at verbose level to help debug this? Would be good to know what the random string is that's received when the server goes offline (likely it's a spurious boot message sent over RS485).

exciton committed:
[modbus] Use uint16_t for frame length to allow full 256 byte frame.

This fixes a bug where the parser is searching for a custom command byte by byte and the frame_length overflows from 255 to 0, thereby searching forever.

exciton committed:
[modbus] Deprecate ready_for_immediate_send. Modbus uses a queue now.

Refactor growat_solar (which was the only component to call ready_for_immedate_send to just queue a command on update() and don't do all the custom tracking of send/waiting, which is not needed. The response will only be returned to the component if it was immediately requested (modbus handles this now).

exciton committed:
[modbus] Fix requeueing logic to only requeue once

exciton commented:
All features of this prototype branch are now merged (with much better code) into dev. You should see everything in 2026.9.0.
